src/gretel_trainer/relational/sdk_extras.py

```diff
@@ -18,90 +18,92 @@
-    try:
-        with open(out_path, "wb+") as out:
-            out.write(smart_open.open(download_link, "rb").read())
-    except:
-        logger.warning(f"Failed to download `{artifact_name}`")
-
-
-def download_tar_artifact(project: Project, artifact_name: str, out_path: Path) -> None:
-    download_link = project.get_artifact_link(artifact_name)
-    try:
-        response = requests.get(download_link)
-        if response.status_code == 200:
-            with open(out_path, "wb") as out:
-                out.write(response.content)
-    except:
-        logger.warning(f"Failed to download `{artifact_name}`")
-
-
-def sqs_score_from_full_report(report: Dict[str, Any]) -> Optional[int]:
-    with suppress(KeyError):
-        for field_dict in report["summary"]:
-            if field_dict["field"] == "synthetic_data_quality_score":
-                return field_dict["value"]
-
-
-def get_record_handler_data(record_handler: RecordHandler) -> pd.DataFrame:
-    return pd.read_csv(record_handler.get_artifact_link("data"), compression="gzip")
-
-
-def start_job_if_possible(
-    job: Job,
-    table_name: str,
-    action: str,
-    project: Project,
-    number_of_artifacts: int,
-) -> None:
-    if job.data_source is None or room_in_project(project, number_of_artifacts):
-        _log_start(table_name, action)
-        job.submit_cloud()
-    else:
-        _log_waiting(table_name, action)
-
-
-def _log_start(table_name: str, action: str) -> None:
-    logger.info(f"Starting {action} for `{table_name}`.")
-
-
-def _log_waiting(table_name: str, action: str) -> None:
-    logger.info(
-        f"Maximum concurrent relational jobs reached. Deferring start of `{table_name}` {action}."
-    )
+class ExtendedGretelSDK:
+    def __init__(self, hybrid: bool):
+        self._hybrid = hybrid
+
+    def get_job_id(self, job: Job) -> Optional[str]:
+        if isinstance(job, Model):
+            return job.model_id
+        elif isinstance(job, RecordHandler):
+            return job.record_id
+        else:
+            raise MultiTableException("Unexpected job object")
+
+    def delete_data_source(self, project: Project, job: Job) -> None:
+        if not self._hybrid and job.data_source is not None:
+            project.delete_artifact(job.data_source)
+
+    def cautiously_refresh_status(
+        self, job: Job, key: str, refresh_attempts: Dict[str, int]
+    ) -> Status:
+        try:
+            job.refresh()
+            refresh_attempts[key] = 0
+        except:
+            refresh_attempts[key] = refresh_attempts[key] + 1
+
+        return job.status
+
+    def download_file_artifact(
+        self,
+        gretel_object: Union[Project, Model],
+        artifact_name: str,
+        out_path: Union[str, Path],
+    ) -> None:
+        download_link = gretel_object.get_artifact_link(artifact_name)
+        try:
+            with open(out_path, "wb+") as out:
+                out.write(smart_open.open(download_link, "rb").read())
```

Comment:
May want to consider using `shutil.copy()` or something to avoid memory bloat on large artifacts, see: https://github.com/Gretellabs/monogretel/pull/2036

Reply:
I don't think we can use `shutil` because `download_link` is not a local path—it's either a cloudfront URL (cloud) or an S3/etc. URL (hybrid).

We may want to consider adding first class support for downloading individual Gretel artifacts to the client. Currently we have `download_artifacts` on Jobs, but of course that downloads _all_ artifacts and we only need one or two. There is no download method for project artifacts, only `get_artifact_link`.

If we did add new `download_artifact` (singular) methods to Projects and Jobs in the client, we could use `requests` with the Cloudfront URL in the cloud context to avoid reading the artifact, but I think in hybrid we'd still be stuck using `smart_open` to handle abstracting over the specific cloud provider object storage being used.

Reply:
But `smart_open` gives you the file-like object that `shutil.copyfileobj` should be able to use.

Reply:
🤯